### ISSUE TITLE
Add annotations to mark asset path and package name

### ISSUE THREAD

**Issue by scheglov** [flutter/flutter#144937]
### Use case

See https://github.com/dart-lang/sdk/issues/46542 for details. In short it would be useful to have code completion inside the `Image.asset()` constructor invocation, for the `name` argument. Taking the optional `package` into account, if I understand the API right.

### Proposal

But because `Image.asset()` might be not the only API that accept assets, at very least there might be user widgets that accept asset names, we need a way to give the Dart analysis server a hint that this `String` is actually an asset name.

So, I propose adding two annotations to `flutter/lib/src/services/asset_manifest.dart` (let me know if there is a better location).

```dart
/// The marker that denotes that the annotated `String` typed formal parameter
/// is the name of an asset.
const Object assetName = Object();

/// The marker that denotes that the annotated `String` typed formal parameter
/// is the name of a package that contains the asset specified with [assetName].
const Object assetPackageName = Object();
```

And `Image.asset()` will be updated:
```dart
  Image.asset(
    @assetName
    String name, {
    super.key,
    AssetBundle? bundle,
<cut>
    @assetPackageName
    String? package,
    this.filterQuality = FilterQuality.low,
    int? cacheWidth,
    int? cacheHeight,
  })
```

**Comment by jonahwilliams** [comment]
If all we need to do is add some annotations, then this seems like a nice win? There is also https://api.flutter.dev/flutter/dart-ui/FragmentProgram/fromAsset.html , but that is a) used less and b) defined in dart:ui

**Comment by goderbauer** [comment]
How would this roughly work? Would the analyzer add special support to recognize this Flutter-defined annotation? Is there a way to make this a little more generic and useful for other packages to plug into it?

**Comment by scheglov** [comment]
Yes, the analyzer will have special knowledge about these Flutter annotations.
I have a few ideas that make me lean toward Flutter-specific implementation.
1. We need to know also about `@assetPackageName String? package`, to look only in this package.
2. We need to look, which assets are specified in `pubspec.yaml`, in `flutter: assets:`.
3. I don't know other use cases, so it is hard to generalize.

**Comment by goderbauer** [comment]
Makes sense. Adding this seems like a win.
